src/java/org/apache/cassandra/db/memtable/pmem/PmemPartition.java

```diff
@@ -626,6 +616,158 @@ public Iterator<RangeTombstone> getMergedRangeTombstones(PartitionUpdate update)
+        final ArrayList<byte[]> artEntryKeyArray = new ArrayList<>();
+        while (rowItr.hasNext())
+        {
+            LongART.Entry nextEntry = rowItr.next();
+            Unfiltered unfiltered = getUnfiltered(nextEntry, true);
+            if (unfiltered != null && isExpired(pmemTableInfo, unfiltered))
+            {
+                artEntryKeyArray.add(nextEntry.getKey());
+            }
+        }
+        if (!artEntryKeyArray.isEmpty())
+        {
+            for (byte[] artEntryKey : artEntryKeyArray)
+            {
+                pmemRowMap.getRowMapTree().remove(artEntryKey, (Long rowhandle) -> {
+                    TransactionalMemoryBlock rowMemoryBlock = heap.memoryBlockFromHandle(rowhandle);
+                    rowMemoryBlock.free();
+                });
+            }
+        }
+
+        return artEntryKeyArray.size();
+    }
+
+    //Removes RangeTombstones which are older then gcgraceseconds
+    private int removeExpiredRangeTombstone()
+    {
+        Iterator<LongART.Entry> tombstoneMarkerItr = pmemRtmMap.getRangeTombstoneMarkerTree().getEntryIterator();
+        final ArrayList<byte[]> artEntryKeyArray = new ArrayList<>();
+        while (tombstoneMarkerItr.hasNext())
+        {
+            LongART.Entry nextEntry = tombstoneMarkerItr.next();
+            Unfiltered unfiltered = getUnfiltered(nextEntry, false);
+            if (unfiltered != null && isExpired(pmemTableInfo, unfiltered))
+            {
+                artEntryKeyArray.add(nextEntry.getKey());
+            }
+        }
+        if (!artEntryKeyArray.isEmpty())
+        {
+            for (byte[] artEntryKey : artEntryKeyArray)
+            {
+                pmemRtmMap.getRangeTombstoneMarkerTree().remove(artEntryKey, (Long rowhandle) -> {
+                    TransactionalMemoryBlock rowMemoryBlock = heap.memoryBlockFromHandle(rowhandle);
+                    rowMemoryBlock.free();
+                });
+            }
+        }
+
+        return artEntryKeyArray.size();
+    }
+
+    private static int getGcBefore(TableMetadata metadata, int nowInSec)
+    {
+        ColumnFamilyStore cfs = Keyspace.open(metadata.keyspace).getColumnFamilyStore(metadata.id);
+        return cfs.gcBefore(nowInSec);
+    }
+
+    // Returns RTM deletion time
+    private static DeletionTime rtmDeletionTime(RangeTombstoneMarker marker)
+    {
+        return marker.isOpen(false) ? marker.openDeletionTime(false) : marker.closeDeletionTime(false);
+    }
+
+    private Unfiltered getUnfiltered(LongART.Entry nextEntry , boolean isRow)
+    {
+        Row.Builder builder = BTreeRow.sortedBuilder();
+        ByteComparable clusteringByteComparable = ByteComparable.fixedLength(nextEntry.getKey());
+        if(isRow)
+        {
+            Clustering<?> clustering = tableMetadata.comparator.clusteringFromByteComparable(ByteArrayAccessor.instance, clusteringByteComparable);
+            builder.newRow(clustering);
+        }
```

Comment:
Should this path really continue with reading a tombstone? Could you explain in a code comment if so?